dateutil/parser/isoparser.py

```diff
@@ -0,0 +1,380 @@
+# -*- coding: utf-8 -*-
+"""
+This module offers a parser for ISO-8601 strings
+
+It is intended to support all valid date, time and datetime formats per the
+ISO-8601 specification.
+"""
+from datetime import datetime, timedelta, time, date
+import calendar
+from dateutil import tz
+
+from functools import wraps
+
+import re
+import six
+
+__all__ = ["isoparse", "isoparser"]
+
+
+def _takes_ascii(f):
+    @wraps(f)
+    def func(self, str_in, *args, **kwargs):
+        # If it's a stream, read the whole thing
+        str_in = getattr(str_in, 'read', lambda: str_in)()
+
+        # If it's unicode, turn it into bytes, since ISO-8601 only covers ASCII
+        if isinstance(str_in, six.text_type):
+            # ASCII is the same in UTF-8
+            try:
+                str_in = str_in.encode('ascii')
+            except UnicodeEncodeError as e:
+                msg = 'ISO-8601 strings contain only ASCII characters'
+                six.raise_from(ValueError(msg), e)
+
+        return f(self, str_in, *args, **kwargs)
+
+    return func
+
+
+class isoparser(object):
+    def __init__(self, sep='T'):
+        """
+        :param sep:
+            A single character that separates date and time portions
+        """
+        if (len(sep) != 1 or ord(sep) >= 128 or sep in '0123456789'):
+            raise ValueError('Separator must be a single, non-numeric '
+                             'ASCII character')
+
+        self._sep = sep.encode('ascii')
```

Comment:
What if sep is a bytestring? In py 2 it'll technically work but it is quite not correct

Reply:
Good point. Adding an xfail test for this.